File: Algorithms/QuickSort.py
def partition1(arr):
    #LOW MEETS HIGH
    #pivot is first element
    #low start at 0 at the hole, high at len(arr) move first
    pivot = arr[0]
    l= 0
    h = len(arr) -1 


def partition2(arr,low,high):
    #CHASE TO PIVOT AT THE END
    #low, loop indices chase till meet pivot at the end
    #assume high is just right before pivot
    print("Partition ",low,high,arr[low:high+1],arr)
    l = low - 1
    pivot = arr[high]
    for loop in range(low,high,1):
        # Try to move low element to front, no matter how low the arr[low] is
        
        if(arr[loop] <= pivot):
            print("Swap")
            l += 1
            arr[l],arr[loop] = arr[loop],arr[l] 
        print(l,loop,high,arr[low:high+1],arr)
    
    # Low elements are at front now , so low + 1 is the middle point where the pivot should stand
    # This point is sorted
    # Swap pivot and the arr[low+1]
    arr[low + 1] , arr[high] = arr[high],arr[low+1]
    

    return low + 1

def quicksort(arr,low,high):
    if high - low > 2:
        s = partition2(arr,low,high)
        quicksort(arr,low,s)
        quicksort(arr,s+1,high)

if __name__ == "__main__":
    arr = [10,5,6,4,7,1,8,3]
    quicksort(arr,0,len(arr)-1)
    print(arr)





        
    
    
            

File: Algorithms/QuickSort.c
#include <stdio.h>
#include <ctype.h>
int partition_with_method_of_chasing_endPivot(int arr[], int low, int high);
void quicksort(int arr[], int low, int high);
void swap(int * a, int *b);
void display_array(int arr[],int size);

int main(int argc, char ** args){
    int arr[] = {7,2,49,2,734,2,3,0,2,4,6};
    int size = sizeof(arr)/sizeof(arr[0]);
    quicksort(arr,0,size-1);
    display_array(arr,size);
    return 0;
}


int partition_with_method_of_chasing_endPivot(int arr[], int low, int high){
    int lo  = low - 1;
    int pivot = arr[high];
    for(int i = 0; i < high ; i++){
        if(arr[i] <= pivot){
            lo++;
            swap(&arr[lo],&arr[i]); // try to move smaller-than-pivot items to front, low is an anchor
        }
    }
    swap(&arr[lo+1],&arr[high]);  // now smaller-than-pivot items is from low to lo , so pivot is lo + 1, put pivot at lo +1 
    return lo + 1;  // lo + 1 is the pivot position, we need it for next partitions
}

void quicksort(int arr[], int low, int high){
    if(high-low>2){
        int p = partition_with_method_of_chasing_endPivot(arr,low,high);
        quicksort(arr,low,p-1);
        quicksort(arr,p+1,high);
    }
}

void display_array(int arr[],int size){
    printf("[");
    for(int i = 0; i < size; i++){
        printf("%d,",arr[i]);
    }
    printf("]\n");
}

void swap(int * a, int * b){
    int temp = *a;
    *a = *b;
    *b = temp;
}




